src/Razor/src/Microsoft.CodeAnalysis.Razor.CohostingShared/Diagnostics/CohostDocumentPullDiagnosticsEndpointBase.cs

```diff
@@ -0,0 +1,130 @@
+
+        // Diagnostics is a little different, because Roslyn is not designed to run diagnostics in OOP. Their system will transition to OOP
+        // as it needs, but we have to start here in devenv. This is not as big a problem as it sounds, specifically for diagnostics, because
+        // we only need to tell Roslyn the document we need diagnostics for. If we had to map positions or ranges etc. it would be worse
+        // because we'd have to transition to our OOP to find out that info, then back here to get the diagnostics, then back to OOP to process.
+        _logger.LogDebug($"Getting diagnostics for {razorDocument.FilePath}");
+
+        var csharpTask = GetCSharpDiagnosticsAsync(razorDocument, correlationId, cancellationToken);
+        var htmlTask = SupportsHtmlDiagnostics
+            ? GetHtmlDiagnosticsAsync(razorDocument, correlationId, cancellationToken)
+            : SpecializedTasks.EmptyArray<LspDiagnostic>();
+
+        try
+        {
+            await Task.WhenAll(htmlTask, csharpTask).ConfigureAwait(false);
+        }
+        catch (Exception e) when (e is not OperationCanceledException)
+        {
+            _logger.LogError(e, $"Exception thrown in PullDiagnostic delegation");
+            throw;
+        }
+
+        var csharpDiagnostics = csharpTask.VerifyCompleted();
+        var htmlDiagnostics = htmlTask.VerifyCompleted();
+
+        _logger.LogDebug($"Calling OOP with the {csharpDiagnostics.Length} C# and {htmlDiagnostics.Length} Html diagnostics");
+        var diagnostics = await _remoteServiceInvoker.TryInvokeAsync<IRemoteDiagnosticsService, ImmutableArray<LspDiagnostic>>(
+            razorDocument.Project.Solution,
+            (service, solutionInfo, cancellationToken) => service.GetDiagnosticsAsync(solutionInfo, razorDocument.Id, csharpDiagnostics, htmlDiagnostics, cancellationToken),
+            cancellationToken).ConfigureAwait(false);
+
+        if (diagnostics.IsDefaultOrEmpty)
+        {
+            return null;
+        }
+
+        _logger.LogDebug($"Reporting {diagnostics.Length} diagnostics back to the client");
+        return diagnostics.ToArray();
+    }
+
+    private async Task<LspDiagnostic[]> GetCSharpDiagnosticsAsync(TextDocument razorDocument, Guid correletionId, CancellationToken cancellationToken)
+    {
+        if (!razorDocument.TryComputeHintNameFromRazorDocument(out var hintName) ||
+            await razorDocument.Project.TryGetSourceGeneratedDocumentFromHintNameAsync(hintName, cancellationToken).ConfigureAwait(false) is not { } generatedDocument)
+        {
+            return [];
+        }
+
+        _logger.LogDebug($"Getting C# diagnostics for {generatedDocument.FilePath}");
+
+        using var _ = _telemetryReporter.TrackLspRequest(LspMethodName, "Razor.ExternalAccess", TelemetryThresholds.DiagnosticsSubLSPTelemetryThreshold, correletionId);
```

Comment:
Assuming `Constants.ExternalAccessServerName` is available here:

```suggestion
        using var _ = _telemetryReporter.TrackLspRequest(LspMethodName, Constants.ExternalAccessServerName, TelemetryThresholds.DiagnosticsSubLSPTelemetryThreshold, correletionId);
```

Reply:
It's not available, and it didn't seem important enough to move.

Reply:
No worries! If you ever end up spending hours scratching your head to figure out why a telemetry query isn't returning all results because this string changed, let me know! 😀😀😀😀😀